components/validator/challenger.go

```diff
@@ -24,232 +32,490 @@ type ProofFetcher interface {
 	}
 
-	if isInProgress {
-		isRelated, err := c.IsRelatedChallenge()
-		if err != nil {
-			return nil, fmt.Errorf("unable to check relationship with challenge: %w", err)
-		}
+	toBlock := new(big.Int).SetUint64(status.CurrentL1.Number)
+	// TODO(0xHansLee): add L1BlockTime to rollup config and change to use it
+	finalizationStartL1Block := new(big.Int).Sub(toBlock, new(big.Int).Div(c.finalizationPeriodSeconds, big.NewInt(12)))
+	// The fromBlock is the maximum value of either genesis block(1) or the first block of the finalization window
+	fromBlock := math.BigMax(common.Big1, finalizationStartL1Block)
+
+	outputSubmittedEvent := c.l2ooABI.Events[KeyEventOutputSubmitted]
+	challengeCreatedEvent := c.colosseumABI.Events[KeyEventChallengeCreated]
+
+	addresses := []common.Address{c.cfg.ColosseumAddr}
+	topics := []common.Hash{challengeCreatedEvent.ID}
+
+	// scan OutputSubmittedEvents only when challenger mode is on
+	if !c.cfg.ChallengerDisabled {
+		addresses = append(addresses, c.cfg.L2OutputOracleAddr)
+		topics = append(topics, outputSubmittedEvent.ID)
+	}
 
-		if !isRelated {
-			c.log.Info("another challenge is in progress")
-			return nil, nil
+	query := ethereum.FilterQuery{
+		FromBlock: fromBlock,
+		ToBlock:   toBlock,
+		Addresses: addresses,
+		Topics:    [][]common.Hash{topics},
+	}
+
+	logs, err := c.l1Client.FilterLogs(c.ctx, query)
+	if err != nil {
+		return fmt.Errorf("failed to get event logs related to outputs: %w", err)
+	}
+
+	for _, vLog := range logs {
+		// for OutputSubmitted event
+		if !c.cfg.ChallengerDisabled && vLog.Address == c.cfg.L2OutputOracleAddr {
+			ev := NewOutputSubmittedEvent(vLog)
+			// handle output
+			c.wg.Add(1)
+			go c.handleOutput(ev.OutputIndex)
+			continue
 		}
 
-		status, err := c.GetStatusInProgress()
-		if err != nil {
-			return nil, fmt.Errorf("unable to get challenge status: %w", err)
+		// for ChallengeCreated event
+		if vLog.Address == c.cfg.ColosseumAddr {
+			// handle challenge
+			ev := NewChallengeCreatedEvent(vLog)
+			if ev.OutputIndex.Sign() == 1 && c.isRelatedChallenge(ev.Asserter, ev.Challenger) {
+				c.log.Info("start to handle challenge", "outputIndex", ev.OutputIndex)
+				c.wg.Add(1)
+				go c.handleChallenge(ev.OutputIndex)
+			}
 		}
+	}
 
-		if !c.cfg.OutputSubmitterDisabled {
-			switch status {
-			case chal.StatusAsserterTurn:
-				return c.Bisect()
-			case chal.StatusChallengerTimeout:
-				challengeId, err := c.GetChallengeInProgress()
-				if err != nil {
-					return nil, err
-				}
+	return nil
+}
 
-				// TODO(pangssu): Is it necessary to submit challengerTimeout transaction?
-				c.log.Info("challenger timed out", "challengeId", challengeId)
-				return nil, nil
+// subscribeL2OutputSubmitted subscribes the OutputSubmitted event from L2OutputOracle contract.
+// If the L2 output root is invalid, create challenge.
+// This function should be called only when challenger mode is on.
+func (c *Challenger) subscribeL2OutputSubmitted() {
+	c.wg.Done()
```

Comment:
I think this part should be called by the defer routine.

Reply:
Oh, mistake. Thanks!